### PR TITLE
[SPARK-25758][ML] Deprecate computeCost on BisectingKMeans

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

The PR proposes to deprecate the `computeCost` method on `BisectingKMeans` in favor of the adoption of `ClusteringEvaluator` in order to evaluate the clustering.

## How was this patch tested?

NA

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @holdenk @srowen

**Review by srowen — COMMENTED** [review]
I see, this is a little more than just deprecating a few methods. Is this basically the same change as for KMeans?

File: python/pyspark/ml/clustering.py
Comment (mgaido91) [reply python/pyspark/ml/clustering.py]: sorry, this was not intended, I am fixing this.

**Review by mgaido91 — COMMENTED** [review]
yes, this is the same change which was done there, ie. deprecate the computeCost, but offer the cost on the training dataset in the summary, in order for the users to be able to still get it.

In addition, I also edited the examples so that they don't include deprecated methods.

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: I'm wondering if this PR is a blocker for Spark 2.4. According to JIRA desciption (Improvement/Minor), we cannot remove this 3.0.0 because we didn't announce deprecation before 3.0.0.

cc @cloud-fan since he is a release manager.

Comment (cloud-fan) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: It looks reasonable to me to deprecate it in 2.4 so that we can remove it in 3.0, if this is the last one. Then we can have a consistent ML API in 3.0 after removing these deprecated APIs.

Comment (dongjoon-hyun) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: Thank you for the decision, @cloud-fan !

File: examples/src/main/java/org/apache/spark/examples/ml/JavaBisectingKMeansExample.java
Comment (dongjoon-hyun) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaBisectingKMeansExample.java]: @mgaido91 .
If we are going to change all `ml` examples for deprecation, we had better change the following, too. 
- https://github.com/apache/spark/blob/master/examples/src/main/python/ml/bisecting_k_means_example.py#L45
```scala
    # Evaluate clustering.
    cost = model.computeCost(dataset)
```

**Comment by dongjoon-hyun** [comment]
@mgaido91 . If you don't mind, could you split this PR into two PRs? One is adding `deprecation` annotation only. The other is adding new API and updating all examples?

File: mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/clustering/BisectingKMeans.scala]: yes this is the last one.

> Then we can have a consistent ML API in 3.0 after removing these deprecated APIs.

Yes, that's my goal in targeting this for 2.4.

Thanks.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaBisectingKMeansExample.java
Comment (mgaido91) [reply examples/src/main/java/org/apache/spark/examples/ml/JavaBisectingKMeansExample.java]: thanks I'll do @dongjoon-hyun

**Comment by mgaido91** [comment]
@dongjoon-hyun sure, thanks. I'll update asap. Thanks.

**Comment by cloud-fan** [comment]
LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by WeichenXu123** [comment]
LGTM. thanks!

**Comment by dongjoon-hyun** [comment]
Merged to master/branch-2.4.

**Comment by dongjoon-hyun** [comment]
Thank you, @mgaido91 and all!

**Comment by holdenk** [comment]
I'm seeing this linked from https://github.com/apache/spark/pull/22764 and I'm wondering if we need to revert this. If the information is not actually available where we tell folks it is I think we need to revert this especially since we are in the middle of the release process. Or raise SPARK-25765 to blocker release blocker.

Have I misunderstood the situation here?

**Comment by dongjoon-hyun** [comment]
I also understand today's situation and agree with @holdenk 's thought about SPARK-25765 as a blocker. Ping @cloud-fan since you are a release manager. How can we proceed SPARK-25765?

Maybe, this is due to `Preparing Spark release v2.4.0-rc4` which happen two hours ago. We are in the middle of unstable situation.

Also, cc @gatorsmile .

**Comment by gatorsmile** [comment]
cc @mengxr WDYT? It does not sound a blocker to me.

**Comment by mengxr** [comment]
We have to revert this PR in branch-2.4. It is not a blocker and we shouldn't merge it to branch-2.4 this late in this already delayed release.

**Comment by gatorsmile** [comment]
Let me revert it. Thanks!

**Comment by gatorsmile** [comment]
Done

**Comment by cloud-fan** [comment]
shall we revert it from master as well? At least we need to update the message `This method is deprecated and will be removed in 3.0.0.`

**Comment by cloud-fan** [comment]
reverted from master. Let's move the discussion to https://github.com/apache/spark/pull/22764

**Pull request closed without merging**
